syntax = "proto3";

import "google/protobuf/any.proto";
import "error.proto";

package bess.pb;

// TODO: add something like PortCommandRequest

// Added to suppress "unused" warning.
message _DummyMessage {
  google.protobuf.Any _dummy = 1;
}

message CreatePortRequest {
  string name = 1;
  string driver = 2;
  uint64 num_inc_q = 3;
  uint64 num_out_q = 4;
  uint64 size_inc_q = 5;
  uint64 size_out_q = 6;
  string mac_addr = 7;
  oneof arg {
    PCAPPortArg pcap_arg = 8;
    PMDPortArg pmd_arg = 9;
    UnixSocketPortArg socket_arg = 10;
    ZeroCopyVPortArg zcvport_arg = 11;
    VPortArg vport_arg = 12;
  }
}

message CreatePortResponse {
  Error error = 1;
  string name = 2;
}

message PCAPPortArg {
  string dev = 1;
}

message PMDPortArg {
  bool loopback = 1;
  uint32 port_id = 2;
  string pci = 3;
  string vdev = 4;
}

message UnixSocketPortArg {
  string path = 1;
}

message ZeroCopyVPortArg {

}

message VPortArg {
  string ifname = 1;
  string docker = 2;
  int64 container_pid = 3;
  string netns = 4;
  repeated int64 rxq_cpus = 5;
  uint64 tx_tci = 6;
  uint64 tx_outer_tci = 7;
  bool loopback = 8;
  repeated string ip_addrs = 9;
}
